embassy-rp/src/pio/mod.rs

```diff
@@ -740,6 +764,100 @@ impl<'d, PIO: Instance + 'd, const SM: usize> StateMachine<'d, PIO, SM> {
+                match (
+                    if config.in_count > 0 {
+                        Some(config.pins.in_base)
+                    } else {
+                        None
+                    },
+                    if config.pins.sideset_count > 0 {
+                        Some(config.pins.sideset_base)
+                    } else {
+                        None
+                    },
+                    if config.pins.set_count > 0 {
+                        Some(config.pins.set_base)
+                    } else {
+                        None
+                    },
+                    if config.pins.out_count > 0 {
+                        Some(config.pins.out_base)
+                    } else {
+                        None
+                    },
+                ) {
+                    (None, None, None, None) => false,
+
+                    (Some(..32), None, None, None) => false,
+                    (None, Some(..32), None, None) => false,
+                    (None, None, Some(..32), None) => false,
+                    (None, None, None, Some(..32)) => false,
+
+                    (Some(..32), Some(..32), None, None) => false,
+                    (None, Some(..32), Some(..32), None) => false,
+                    (None, None, Some(..32), Some(..32)) => false,
+                    (Some(..32), None, None, Some(..32)) => false,
+
+                    (None, Some(..32), Some(..32), Some(..32)) => false,
+                    (Some(..32), None, Some(..32), Some(..32)) => false,
+                    (Some(..32), Some(..32), None, Some(..32)) => false,
+                    (Some(..32), Some(..32), Some(..32), None) => false,
+
+                    (Some(..32), Some(..32), Some(..32), Some(..32)) => false,
+
+                    (Some(16..), None, None, None) => true,
+                    (None, Some(16..), None, None) => true,
+                    (None, None, Some(16..), None) => true,
+                    (None, None, None, Some(16..)) => true,
+
+                    (Some(16..), Some(16..), None, None) => true,
+                    (None, Some(16..), Some(16..), None) => true,
+                    (None, None, Some(16..), Some(16..)) => true,
+                    (Some(16..), None, None, Some(16..)) => true,
+
+                    (None, Some(16..), Some(16..), Some(16..)) => true,
+                    (Some(16..), None, Some(16..), Some(16..)) => true,
+                    (Some(16..), Some(16..), None, Some(16..)) => true,
+                    (Some(16..), Some(16..), Some(16..), None) => true,
+
+                    (Some(16..), Some(16..), Some(16..), Some(16..)) => true,
+
+                    (i, side, set, out) => panic!(
+                        "All pins must either be <=31 or >=16, in:{:?}, side:{:?}, set:{:?}, out:{:?}",
+                        i, side, set, out
+                    ),
+                }
+            };
+
+            if shift_gpio_base {
+                sm.pinctrl().write(|w| {
```

Comment:
instead of duplicating this you can make a `shift` variable that's either `0` or `16` and substract it.